persistence/remote/src/main/java/org/infinispan/persistence/remote/internal/GlobalRemoteContainersImpl.java

```diff
@@ -0,0 +1,89 @@
+package org.infinispan.persistence.remote.internal;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.client.hotrod.configuration.ConfigurationBuilder;
+import org.infinispan.commons.marshall.Marshaller;
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.infinispan.configuration.global.GlobalConfiguration;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.factories.annotations.Start;
+import org.infinispan.factories.annotations.Stop;
+import org.infinispan.factories.scopes.Scope;
+import org.infinispan.factories.scopes.Scopes;
+import org.infinispan.persistence.remote.configuration.global.RemoteContainerConfiguration;
+import org.infinispan.persistence.remote.configuration.global.RemoteContainersConfiguration;
+import org.infinispan.persistence.remote.global.GlobalRemoteContainers;
+import org.infinispan.persistence.remote.logging.Log;
+import org.infinispan.util.concurrent.BlockingManager;
+
+/**
+ * @since 15.0
+ **/
+@Scope(Scopes.GLOBAL)
+public class GlobalRemoteContainersImpl implements GlobalRemoteContainers {
+
+   @Inject
+   GlobalConfiguration globalConfiguration;
+
+   @Inject
+   BlockingManager blockingManager;
+
+   private final Map<String, CompletionStage<RemoteCacheManager>> remoteContainers = new ConcurrentHashMap<>();
+
+   @Start
+   public void start() {
+      RemoteContainersConfiguration configuration = globalConfiguration.module(RemoteContainersConfiguration.class);
+      if (configuration != null) {
+         for (Map.Entry<String, RemoteContainerConfiguration> e : configuration.configurations().entrySet()) {
+            remoteContainers.put(e.getKey(), CompletableFutures.completedNull());
+         }
+      }
+   }
+
+   @Stop
+   public void stop() {
+      remoteContainers.values().forEach(c -> c.thenAccept(this::stop));
+   }
+
+   private void stop(RemoteCacheManager rcm) {
+      if (rcm != null) rcm.stop();
+   }
+
+   @Override
+   public CompletionStage<RemoteCacheManager> cacheContainer(String name, Marshaller marshaller) {
+      CompletionStage<RemoteCacheManager> cs = remoteContainers.computeIfPresent(name, (k, cf) ->
+            cf.thenCompose(prev -> prev != null
+                  ? CompletableFuture.completedFuture(prev)
+                  : createCacheManager(k, marshaller)));
+      // The entry always exists.
```

Comment:
Is this true? Can't the RemoteStore have the wrong name configured? I think we should check for null and log an ERROR if so.

Reply:
Updated.